Update to Minio Bucket Notification Guide

jamesneimanconsulting opened the pull request:
initial edits

- [X] I have updated the documentation accordingly.

jamesneimanconsulting committed:
Update to Minio Bucket Notification Guide

initial edits

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -1,54 +1,71 @@
-An example configuration for RabbitMQ is shown below:
+| `enable` | `bool` | (_Required_) Specifies if this server endpoint configuration is active/enabled. |
+| `url` | `string` | (_Required_) The AMQP server endpoint (e.g. `amqp://myuser:mypassword@localhost:5672`). |
+| `exchange` | `string` | The name of the exchange. |
```
@harshavardhana do we need to list the various exchange types here?

harshavardhana replied:
yes we need we haven't done this in the past, but anyone configuring amqp already knows this.

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -1,54 +1,71 @@
+| `routingKey` | `string` | The routing key for publishing. |
+| `exchangeType` | `string` | The type of exchange. |
+| `deliveryMode` | `uint8` | The delivery mode for publishing. Set this field to 0 or 1 for transient, or 2 for persistent. |
+| `mandatory` | `bool` | A flag related to publishing. |
```
"related to" is too ambiguous here, related how?

jamesneimanconsulting replied:
@eco-minio There are a number of parameters from the original doc that didn't specify what they are. Can you provide the definitions for these parameters?

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -68,33 +85,44 @@ An example configuration for RabbitMQ is shown below:
 ```
-To update the configuration, use `mc admin config get` command to get the current configuration file for the minio deployment in json format, and save it locally.
+
+Minio supports all of the exchanges available in [RabbitMQ](https://www.rabbitmq.com/). The configuration above uses the `fanout` exchange.
````
Do we need a more direct URL here? (e.g., https://www.rabbitmq.com/tutorials/amqp-concepts.html)

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -68,33 +85,44 @@ An example configuration for RabbitMQ is shown below:
+Restart Minio Server for the changes to take effect. A response similar to this one should be displayed at startup if there were no errors:
 
+```sh
+SQS ARNs:  arn:minio:sqs::1:amqp
````
Might be useful here to show more of the server startup message so it is clear where a user should be looking

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -68,33 +85,44 @@ An example configuration for RabbitMQ is shown below:
+```
 
-### Step 2: Enable bucket notification using Minio client
+**Note:** An arbitrary number of AMQP server endpoint configurations can be added by providing an identifier (e.g. `1` in the example above) for the AMQP instance, and an object name for the per-server configuration parameters.
````
I think this needs a more in depth explanation, it isn't immediately clear that the user essentially needs to create an additional versioned entry in the config.

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -125,46 +153,48 @@ channel.basic_consume(callback,
+| `enable` | `bool` | (_Required_) Specifies if this server endpoint configuration is active/enabled. |
+| `broker` | `string` | (_Required_) The MQTT server endpoint (e.g. `tcp://localhost:1883`). |
+| `topic` | `string` | (_Required_) The name of the MQTT topic on which to publish (e.g. `minio`). |
+| `qos` | `int` | Sets the quality-of-service (QOS) level. |
```
Would be good to include a link to what QOS levels are available

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -179,34 +209,45 @@ An example configuration for MQTT is shown below:
+SQS ARNs: arn:minio:sqs::1:mqtt
+```
 
+Minio Server can communicate with an MGTT server using MQTT 3.1 and MGTT 3.1.1. Minio Server can connect to these servers over TCP, TLS, or a Websocket connection using ``tcp://``, ``tls://``, or ``ws://`` respectively, as the scheme for the broker URL. For more information see the [Go Client Documentation](http://www.eclipse.org/paho/clients/golang/).
````
Is MGTT a typo here?

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -231,54 +272,54 @@ client.connect("localhost",1883,60)
 
-```py
-python mqtt.py
-{“Records”:[{“eventVersion”:”2.0",”eventSource”:”aws:s3",”awsRegion”:”",”eventTime”:”2016–09–08T22:34:38.226Z”,”eventName”:”s3:ObjectCreated:Put”,”userIdentity”:{“principalId”:”minio”},”requestParameters”:{“sourceIPAddress”:”10.1.10.150:44576"},”responseElements”:{},”s3":{“s3SchemaVersion”:”1.0",”configurationId”:”Config”,”bucket”:{“name”:”images”,”ownerIdentity”:{“principalId”:”minio”},”arn”:”ar ...
````
I think this needs to be closed with a triple tick, it is causing weird rendering issues in the json output below and a few other characters on the page

![minio_readme_md_at_6cfa49f9ce099d5b8f818607b4804bf281064e94_ _minio_minio](https://user-images.githubusercontent.com/41090896/49108579-64a10080-f23d-11e8-990c-7e73788a2dcb.png)

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -231,54 +272,54 @@ client.connect("localhost",1883,60)
+### 4.1 Install Elasticsearch
 
-When the _namespace_ format is used, Minio synchronizes objects in the bucket with documents in the index. For each event in the Minio, the server creates a document with the bucket and object name from the event as the document ID. Other details of the event are stored in the body of the document. Thus if an existing object is over-written in Minio, the corresponding document in the Elasticsearc ...
+Minio requires Elasticsearch 5.x which is the latest major release series.
```
AFAIK the latest version of ES is 6.5, we should remove any reference to "latest" as that will not age well.  @harshavardhana did we verify that these instructions work with the latest version?

harshavardhana replied:
No, we cannot guarantee, but anything 5.x is okay.

jamesneimanconsulting committed:
Fixed issue with quote marks

Hopefully these replacements fix the broken code block renderin issues.

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -291,46 +332,60 @@ An example of Elasticsearch configuration is as follows:
 ```
 
-If Elasticsearch has authentication enabled, the credentials can be supplied to Minio via the `url` parameter formatted as `PROTO://USERNAME:PASSWORD@ELASTICSEARCH_HOST:PORT`.
+If authentication is enabled Elasticsearch, the credentials are supplied to Minio via the `url` parameter formatted as `PROTO://USERNAME:PASSWORD@ELASTICSEARCH_HOST:PORT`.
````
"...is enabled in..."

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -68,33 +85,44 @@ An example configuration for RabbitMQ is shown below:
+To update the configuration, use `mc admin config get` to get the current JSON configuration file for the Minio deployment, and save it locally:
+
 ```sh
 $ mc admin config get myminio/ > /tmp/myconfig
````
These instructions are listed multiple times throughout the document.  Is there an elegant way to present them as a  single part of the document? Same thing with `mc admin config set`.

jamesneimanconsulting committed:
Updates to Minio Bucket Notification Guide

Fixed issues causing improperly code highlighting:
- missing commas in JSON
- mc notation using > or <

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -231,54 +272,54 @@ client.connect("localhost",1883,60)
+This notification target supports two formats:
 
-### Step 1: Ensure minimum requirements are met
+* **namespace**: Minio Server synchronizes objects in the bucket with documents in the index. For each event in Minio, the server creates a document, and a document ID composed of the bucket and object name; details of the event are stored in the body of the document. If an existing object is overwritten in Minio, the corresponding document in the Elasticsearch index is updated. If an object is d ...
```
Similar to my comment above about config get/set, is there a way to elegantly put these in a single place in the document?  Only trouble I can see is that namespace / access is only applicable to a handful of the event targets.

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -291,46 +332,60 @@ An example of Elasticsearch configuration is as follows:
 ```
 
-Note that, you can add as many Elasticsearch server endpoint configurations as needed by providing an identifier (like "1" in the example above) for the Elasticsearch instance and an object of per-server configuration parameters.
+Restart Minio Server for the changes to effect. A response similar to this one should be displayed at startup if there were no errors:
````
"...for the changes to take effect..."

harshavardhana commented on docs/bucket/notifications/README.md:
```diff
@@ -1,54 +1,71 @@
+8. [Publish Minio Events via MySQL](#publish-minio-events-via-mysql) 
+9. [Publish Minio Events via Kafka](#publish-minio-events-via-kafka) 
+10. [Publish Minio Events via Webhooks](#publish-minio-events-via-webhooks)
 
```
I guess 1.. 10 naming convention is a bit weird and looks odd to the eye. Would it be possible to simply just leave the numbering and just keep it like 

```
- Topic
- Topic
```
![screenshot from 2018-11-27 15-28-29](https://user-images.githubusercontent.com/622699/49118480-37fae200-f259-11e8-863e-eb00704558d2.png)

Other markdown examples which use the `-` style https://github.com/avelino/awesome-go#contents

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -847,45 +990,74 @@ mc mb myminio/images
 # Print out the notification configuration on the `images` bucket.
 mc events list myminio/images
-arn:minio:sqs::1:postgresql s3:ObjectCreated:*,s3:ObjectRemoved:* Filter: suffix=”.jpg”
+arn:minio:sqs::1:postgresql s3:ObjectCreated:*,s3:ObjectRemoved:* Filter: suffix=".jpg"
```
Change arn:minio:sqs::1:postgresql to arn:minio:sqs::1:mysql

jamesneimanconsulting committed:
fixed highlighting issues

Checked in both GH preview and Gluegun

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -291,46 +332,60 @@ An example of Elasticsearch configuration is as follows:
 ```
 
-Use curl to view contents of ``minio_events`` index.
+Use `curl` to view contents of ``minio_events`` index:
````
Change double ticks to single in minio_events

jamesneimanconsulting committed:
Fixed SQL output

Went back to single line output

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -524,17 +598,24 @@ The NATS configuration block in ``config.json`` is as follows:
+
+Restart Minio server for the changes to take effect. 
+
+**Note:** ``bucketevents`` is the subject used by NATS in this example.
```
Change double ticks to single in `bucketevents`

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -602,28 +690,42 @@ func main() {
 ```
 
-The example ``nats.go`` program prints event notification to console.
+Once the upload completes, return to the first console. A response similar to this one should be displayed:
````
Similar feedback as the config get/set, maybe a section called testing at the end?

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -653,59 +755,73 @@ func main() {
+### 7.2 Add a PostgreSQL Endpoint to Minio
 
-The Minio server configuration file is stored on the backend in json format. The PostgreSQL configuration is located in the `postgresql` key under the `notify` top-level key. Create a configuration key-value pair here for your PostgreSQL instance. The key is a name for your PostgreSQL endpoint, and the value is a collection of key-value parameters described in the table below.
+The Minio Server configuration file is stored on the backend in JSON format. The PostgreSQL configuration is located in the `postgresql` key under the top-level `notify` key. Create a configuration key-value pair here for the PostgreSQL instance. The key is a name for the PostgreSQL endpoint, and the value is a collection of  parameters described in the table below:
```
Same feedback as config get/set

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -821,24 +949,32 @@ An example of MySQL configuration is as follows:
-To update the configuration, use `mc admin config get` command to get the current configuration file for the minio deployment in json format, and save it locally.
-```sh
+
+**Note:** The root password has been set to `password` and the name of the database has been set to `miniodb` for illustrative purposes.
````
might change this to "database root password" to avoid confusion with the OS

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -847,45 +983,53 @@ mc mb myminio/images
+### 9.1 Install Apache Kafka
 
-Minio requires Kafka version 0.10 or 0.9. Internally Minio uses the [Shopify/sarama](https://github.com/Shopify/sarama/) library and so has the same version compatibility as provided by this library.
+Minio requires Kafka version 0.9 or 0.10. Internally, Minio uses the [Shopify/sarama](https://github.com/Shopify/sarama/) library and has the same version compatibility as provided by this library.
```
@harshavardhana is this still the case?

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -894,191 +1038,174 @@ The Minio server configuration file is stored on the backend in json format. Upd
-After updating the NSQ configuration in /tmp/myconfig , use `mc admin config set` command to update the configuration for the deployment.Restart the Minio server to put the changes into effect. The server will print a line like `SQS ARNs: arn:minio:sqs::1:nsq` at start-up if there were no errors.
 ```sh
-$ mc admin config set myminio < /tmp/myconfig
+NODE_ENV=webhook node thumbnail-webhook.js
````
Is this supposed to be an `export`?

jamesneimanconsulting replied:
@eco-minio "NODE_ENV=webhook node thumbnail-webhook" was the command provided to start the server but i couldn't get thumbnailer to work - just says "Error: Cannot find module '.../Minio/gluegun_clone/gluegun/thumbnail'" when i run that command. Can you please suggest a correction or advise in this case?

eco-minio commented on docs/bucket/notifications/README.md:
````diff
@@ -894,191 +1038,174 @@ The Minio server configuration file is stored on the backend in json format. Upd
 ```
 
-Note that, you can add as many NSQ daemon endpoint configurations as needed by providing an identifier (like "1" in the example above) for the NSQ instance and an object of per-server configuration parameters.
+Configure Minio Server to send notifications to this URL (as mentioned in step 1) and use ``mc`` to set up bucket notifications (as mentioned in step 2).
````
Replace double ticks for mc

eco-minio commented on docs/bucket/notifications/README.md:
```diff
@@ -894,191 +1038,174 @@ The Minio server configuration file is stored on the backend in json format. Upd
 
-
-*NOTE* If you are running [distributed Minio](https://docs.minio.io/docs/distributed-minio-quickstart-guide), modify ``~/.minio/config.json`` on all the nodes with your bucket event notification backend configuration.
+**Note:** For [distributed Minio](https://docs.minio.io/docs/distributed-minio-quickstart-guide), modify ``~/.minio/config.json`` on all of the nodes with the backend configuration for bucket event notifications.
```
@harshavardhana this should be removed correct?

jamesneimanconsulting committed:
Updates to Minio Bucket Notification Guide

Revised per reviewer feedback.

harshavardhana committed:
Merge branch 'new-docs' into patch-10
